doc/ext-pake/proposed/spake2plus_proposal_draft.md

```diff
@@ -0,0 +1,67 @@
+Introduction
+------------
+
+This document proposes protocol flow and high-level API changes required in PSA Cryptography API 1.1 PAKE Extension, to support the SPAKE2+ protocol.
+
+SPAKE2+ Version
+---------------
+
+SPAKE2+, an Augmented PAKE Draft 02, 10 December 2020 is considered for proposal.
+
+Link : [https://tools.ietf.org/pdf/draft-bar-cfrg-spake2plus-02.pdf](https://tools.ietf.org/pdf/draft-bar-cfrg-spake2plus-02.pdf)
+
+Remarks
+
+*   SPAKE2+, an Augmented PAKE Draft 08, 5 May 2022 is the latest draft version. Link : [https://datatracker.ietf.org/doc/pdf/draft-bar-cfrg-spake2plus-08](https://datatracker.ietf.org/doc/pdf/draft-bar-cfrg-spake2plus-08)
+*   Shared Secret Key generation is not compatible between Draft 02 and 08.
+*   As most SPAKE2+ implementations e.g. Matter Specification Version 1.0 are based on Draft 02, this version is being considered for better interoperability.
+
+Expected PAKE API Flow :
+------------------------------------
+![](SPAKE2PLUS.svg)
+
+
+### Ciphersuite
+
+The current PAKE ciphersuite does not have encoding for MAC. The SPAKE2+ draft recommends HMAC and CMAC for the MAC operation, therefore we should add MAC field in the ciphersuite.
+### Out of band set up
+
+SPAKE2+ has an offline initialization step where both parties may share parties identities and additional data (the context). This step is considered out of scope but these shared information may be required for SPAKE2+ protocol execution.
+
+**Shared Information** : ProverID, VerifierID and Context.
+
+**Input methods** :
+
+*   Existing API's (when user is Prover)
+    *   `psa_pake_set_user()` to input ProverID
+    *   `psa_pake_set_peer()` to input VerifierID
+    *   `psa_pake_set_role()` to set role as client/server
+    *   `psa_pake_input()` with *(new)* `PSA_PAKE_STEP_ADDITIONAL_DATA` step to input context (additional data).
+
+### Password, PBKDF Output and Verification value (W0, W1, L)
+
+SPAKE2+ requires PBKDF Output and Verification value (W0, W1, L) generated from password as below :
+
+ `w0s || w1s = PBKDF(len(pw) || pw || len(A) || A || len(B) || B)`  
+ `W0 = w0s mod p`  
+ `W1 = w1s mod p`  
+ `L = w1*P`
+
+Computing these values is considered Out of Scope for PAKE API as this a registration phase which could happen out of actual protocol flow.
+
+Since W0 & W1 and L serve as registartion records and verification value, these values can be treated as secrets.
```

Comment:
> So I agree that it would be best if it is possible to optionally input the password instead of the values derived using it. It would also be nice if you didn't have to import the password before usage, similar to `psa_key_derivation_input_bytes()`.

At the moment, when a secret input is provided to a KDF via `psa_key_derivation_input_bytes()`, this prevents the application using `psa_key_derivation_output_key()`. See the documentation for [`PSA_KEY_DERIVATION_INPUT_SECRET`](https://arm-software.github.io/psa-api/crypto/1.1/api/ops/kdf.html#c.PSA_KEY_DERIVATION_INPUT_SECRET), [`PSA_KEY_DERIVATION_INPUT_PASSWORD`](https://arm-software.github.io/psa-api/crypto/1.1/api/ops/kdf.html#c.PSA_KEY_DERIVATION_INPUT_PASSWORD), and [`psa_key_derivation_output_key()`](https://arm-software.github.io/psa-api/crypto/1.1/api/ops/kdf.html#c.psa_key_derivation_output_key).

If we provide a high-level function, such as `psa_pake_registration()`, that outputs a key that can be stored in the key-store, then it _might_ be appropriate to permit a non-key input of the password, because storing a SPAKE2+ password in the keystore is not required, and not recommended. However, I am not sure we should relax the 'secret key input for secret key output' requirement that the PAKE or key derivation operations currently have.

Reply:
> In practice, a concrete PBKDF is required, with specific input parameters (for both the construction of the input keying material, the salt and any iteration count), and a specific length for the `w0s||w1s` output string.
> 
> Matter does define exactly how this part of the protocol must run, including the absence of identity-strings, and how the PBKDF2 parameters are shared and included in the transcript context.
> 
> If we define a higher-level API, such as `psa_pake_registration()`, that matches the Matter requirements, does this become a Matter-specific API? - or do we add enough parameters to support other application uses that do have identities, and generate different lengths of intermediate `w0s||w1s` output?

and 

> At the moment, when a secret input is provided to a KDF via `psa_key_derivation_input_bytes()`, this prevents the application using `psa_key_derivation_output_key()`

I have defined the `psa_pake_registration()` to take as input a PBKDF operation and output key attributes. The PBKDF operation will be initialized OOB and can accept a non-key password as input. We can import the non-key output of the KDF with attributes provided as input and store the key id in the PAKE operation.